### PR TITLE
Add ADR for Testing Flows

### PR DESCRIPTION
**Pull request by kevindew**
Trello: https://trello.com/c/qQL1Uuwh/421-review-and-feedback-on-smart-answers-integration-test-proposal

This ADR aims to document the discussions and decisions that have gone
into the definition of a new testing flow approach for Smart Answers.

👉   [View rendered](https://github.com/alphagov/smart-answers/blob/testing-flows-adr/docs/arch/003-testing-flows.md)

This builds upon two PR's to seek feedback:

- https://github.com/alphagov/smart-answers/pull/5394
- https://github.com/alphagov/smart-answers/pull/5440

### CODE REVIEW

File: docs/arch/003-testing-flows.md [hunk 1]
```diff
@@ -0,0 +1,174 @@
+# ADR 3: Testing Flows
+
+Date: 2021-06-29
+
+## Context
+
+An individual Smart Answer is defined by a Flow, this is a class which defines the questions, outcomes and associated logic for the journey a user takes through a Smart Answer. There are currently 3 different approaches used for testing a Flow and none of these are applied consistently across all of them. This ADR concerns itself with the definition of an approach that can be consolidate this inconsistency.
```
Comment (gclssvglx) [inline docs/arch/003-testing-flows.md:7]: You might have meant: `an approach that can consolidate this inconsistency`

File: docs/arch/003-testing-flows.md [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+- assertions made at an ERB level (to prove their execution)
+- a method to achieve full (or at least very high) coverage of Flow files
+- easy to understand tests (relatively, given how hard Smart Answers can be to understand)
+- relatively low levels of tedium/repetition in reading/writing tests (ideally length of test flow file is within 2x the flow file length)
+- testing approaches that are consistent across different flow approaches (URI path, query string and session)
+
+Following [two](https://github.com/alphagov/smart-answers/pull/5394) [rounds](https://github.com/alphagov/smart-answers/pull/5440) of feedback we've established that the new approach will:
+
+- consolidate integration and unit testing to a single approach, similar to Rails' controller testing (with a similar emphasis on integration testing), since there is high possibility of tedious levels of duplication between multiple files
+- have a boundary of testing the Flow and it's contents, explicitly not testing earlier considerations such as routing or controller - this is due to them being tested separately and the performance advantages gained from not including them
```
Comment (gclssvglx) [inline docs/arch/003-testing-flows.md:50]: You might have meant: `have a boundary of testing the Flow and its contents` as it is possessive.

File: docs/arch/003-testing-flows.md [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+
+An individual Smart Answer is defined by a Flow, this is a class which defines the questions, outcomes and associated logic for the journey a user takes through a Smart Answer. There are currently 3 different approaches used for testing a Flow and none of these are applied consistently across all of them. This ADR concerns itself with the definition of an approach that can be consolidate this inconsistency.
+
+The existing approaches are:
+
+### [Flow integration tests](https://github.com/alphagov/smart-answers/tree/93f34cae3eb746b5b08d78b6ea6b7d5c1dfbbc18/test/integration/smart_answer_flows)
+
+These tests:
+
+- have widely adoption, 31 / 36 Flows have tests
```
Comment (gclssvglx) [inline docs/arch/003-testing-flows.md:15]: You might have meant: `have wide adoption` or `are widely adopted`

File: docs/arch/003-testing-flows.md [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+- not assert directly on ERB output unless there is logic in ERB files (instead asserting they output something to prove no syntax errors), to reduce test verbosity and friction when copy changes
+- test Flows in a style organised around the nodes (questions and outcomes), for understandability of what code is being tested:
+  - start pages will be tested to [assert they have a title and a body](#set-up-and-start-page)
+  - questions will test [all subsequent nodes are reachable](#asserting-a-question-goes-to-a-next-node) and [any validations](#testing-validation), they may be used to test calculators for complex scenarios (in simple scenarios they would already be tested implicitly), however the emphasis is on testing user visible scenarios
+  - outcomes will only be tested explicitly [when there is specific logic for them](#testing-logic-in-an-outcome), they will already be tested implicitly for questions reaching outcomes
+
+
+This approach offers a number of advantages over the existing testing approaches, but it does also arrive with some concerns which are noted for prosperity. These concerns are:
```
Comment (gclssvglx) [inline docs/arch/003-testing-flows.md:58]: I think you might have meant: `posterity` not `prosperity`

**Review by gclssvglx — COMMENTED** [review]
Looking really great 👍   Unfortunately, I got there first to comment on some typos 😉

**Comment by kevindew** [comment]
Awesome, thanks @gclssvglx all great catches 💯

**Review by leenagupte — APPROVED** [review]
All sounds good to me. 👍

**Pull request merged**